src/persist-txn/src/lib.rs

```diff
@@ -479,10 +479,10 @@ pub(crate) async fn empty_caa<S, F, K, V, T, D>(
 {
     let name = name();
     let empty: &[((&K, &V), &T, D)] = &[];
-    let mut upper = txns_or_data_write
-        .shared_upper()
-        .into_option()
-        .expect("shard should not be closed");
+    let Some(mut upper) = txns_or_data_write.shared_upper().into_option() else {
+        // Shard is closed.
```

Comment:
nit: "Shard is closed, which means the upper must be past init_ts" (and similar below)